# Learn v75 med walkthrough-metoden

In [151]:
import pandas as pd 
import numpy as np 
from catboost import CatBoostClassifier,Pool,cv,utils

import sys
sys.path.append('C:\\Users\peter\\Documents\\MyProjects\\PyProj\\Trav\\spel')
import V75_scraping as vs

In [152]:
# df skall innehålla datum,avd,vodds
def proba_order_score(df_, y,proba):
    kassa=1000
    df = df_.copy()
    df['proba'] = proba[:,1]
    df['f'] = (df.proba*df.vodds - 1) / (df.vodds-1)  # kelly formel
    df['spela'] = df.f >0
    df['insats'] = df.spela * df.f * kassa

    df.sort_values(['datum','avd','proba'],ascending=[True,True,False],inplace=True)
    proba_order=df.groupby(['datum','avd']).proba.cumcount()

    df['prob_order']=proba_order+1
    df['y'] = y
    
    return df, df.loc[df.y==1].prob_order.mean()   # mean prob_order för vinnarhäst

In [153]:
def create_ekipage(df_):
    df=df_.copy()
    prefs = ['','h1_','h2_','h3_','h4_','h5_',]
    for pr in prefs:
        df[pr+'ekipage'] = df[pr+'kusk'].str.cat(df['häst'], sep =", ")
        df.drop([pr+'kusk'],axis=1, inplace=True)
        
    return df.drop(['häst'], axis=1)

In [154]:
### returnera en modell med parametrar satta
def get_model(d=6,l2=2,iterations=3000,use_best=True,verbose=False):
    model = CatBoostClassifier(iterations=iterations,use_best_model=use_best, 
        custom_metric=['Logloss', 'AUC','Recall', 'Precision', 'F1', 'Accuracy'],

        eval_metric='Accuracy', 
        depth=d,l2_leaf_reg=l2,
        auto_class_weights='Balanced',verbose=verbose, random_state=2021) 
    return model                

In [293]:
### Features som inte används vid träning
def remove_features(df_,remove_mer=[]):
    # df = df_.copy()
    #remove_mer=['h5_perf','h5_auto','h4_perf','h4_auto', 'h3_perf', 'h2_perf']
    df = df_.drop(['avd','startnr','vodds','podds','bins','h1_dat','h2_dat','h3_dat','h4_dat','h5_dat'],axis=1) #
    if remove_mer:
        df = df.drop(remove_mer,axis=1)
    
    # df=check_unique(df.copy())
    # df=check_corr(df.copy())
    return df

In [156]:
 ## byt ut alla NaN till text för cat_features
def replace_NaN(X_train,X_test=None, cat_features=[]):
    # print('cat_features',cat_features)
    for c in cat_features:
        # print(c)
        X_train.loc[X_train[c].isna(),c] = 'None'       ### byt ut None-värden till texten 'None
        if X_test is not None:  ## om X_test är med
            X_test.loc [X_test[c].isna(),c] = 'None'    ### byt ut None-värden till texten 'None

    return X_train,X_test

In [157]:
def scrape_nya_lopp():
    nya_lopp,strukna = vs.v75_scraping(resultat=True,history=True)

    df=pd.concat([pd.read_csv('all_data.csv'), nya_lopp])
    print('shape med nya lopp',df.shape)
    #ta bort dubletter
    df.drop_duplicates(['datum','avd','häst'],inplace=True)
    df.sort_values(by=['datum','avd'],inplace=True)
    df.reset_index(drop=True,inplace=True)
    print('shape med dubletter bort',df.shape)

    df.to_csv('all_data.csv', index=False)

    print("första datum i df =",df.datum.head(1).to_list()[0])
    print("sista  datum i df =",df.datum.tail(1).to_list()[0])

    return df,nya_lopp

In [158]:
### beräkna vilka datum att använda ###
def get_alla_datum(test_from_proc=0.75, train_from_proc=0, total_omlärning = False):
    if total_omlärning:
        nya_lopp=None
        df = pd.read_csv('all_data.csv')     
        alla_datum = df.datum.unique()
        split_ix = int(len(alla_datum)*test_from_proc)
    else:
        # normalt adderar vi bara 1 eller flera veckor från "omg_att_spela_link.csv"
        df, nya_lopp = scrape_nya_lopp()  # scrape från 'omg_att_spela_link.csv' och addera till df
        omg_df = pd.read_csv('omg_att_spela_link.csv')     
        startix=omg_df.Link.str.find('spel')[0]    # index till 'spel' i url
        alla_datum = omg_df.Link.str.slice(start=startix+5,stop=startix+15).to_list() # en datum 
        split_ix=0
        print(f'datum att lära från {alla_datum}')

    return df,nya_lopp,alla_datum,split_ix


## Walkthrough-funktionen  här

In [159]:

### Kör en walkthrough learn här, en datum i taget framåt

# Jag har ändrat till att alla steg kör utan test-datam ed fast iterations=100
def walkthrough(classic_test=False, verbose=False):
    
    df, nya_lopp, alla_datum, split_ix = get_alla_datum()

    l2_leaf_regs=2
    model=get_model(use_best=False,iterations=100)
    df=remove_features(df.copy())
    cat_features = list(df.loc[:,df.dtypes=='O'].columns)
    df,_ = replace_NaN(df.copy(), cat_features=cat_features)    
    print(f'cat_features {cat_features}\n')

    df['plac']=(df.plac==1)*1
        
    for nr,datum in enumerate(alla_datum[split_ix:]):
        print(f'walk-iter {nr+1} av {len(alla_datum[split_ix:])} ',end=': ')

        X_train = df.loc[df.datum<datum,:].copy()
        y_train = X_train.plac; X_train.drop(['plac'],axis=1,inplace=True)

        if classic_test:    ### klassisk train/test utan walkthrough
            X_test  = df.loc[df.datum>=datum,:].copy()
            y_test  = X_test.plac;  X_test.drop(['plac'],axis=1,inplace=True)
            train_pool = Pool(X_train,y_train,cat_features=cat_features)
            test_pool = Pool(X_test,y_test,cat_features=cat_features)
            model.fit(train_pool,use_best_model=True, verbose=verbose,eval_set=test_pool)
        else:
            X_test  = df.loc[df.datum==datum,:].copy()
            y_test  = X_test.plac;  X_test.drop(['plac'],axis=1,inplace=True)
            train_pool = Pool(X_train,y_train,cat_features=cat_features)
            test_pool = Pool(X_test,y_test,cat_features=cat_features)
            model.fit(train_pool,use_best_model=False, verbose=verbose)

        print('best iteration',model.get_best_iteration(), '\tbest score', round(model.get_best_score()['learn']['Accuracy'],3) )
        ##['validation']['Logloss'],3),'\t', round(model.get_best_score()['validation']['Accuracy:use_weights=true'],3))
        
        if classic_test:    ### klassisk train/test utan walkthrough
            return model,cat_features
    
        model.save_model('modeller/model_'+datum)

    X_train =df.copy().drop('plac',axis=1)
    y_train = df.plac 
    model.fit(X_train,y=y_train,cat_features=cat_features)
    print(f'spara model_senaste',datum)
    model.save_model('modeller/model_senaste')

    return df,nya_lopp, model,cat_features

### Här körs hela walkthrough

In [160]:
df, nya_lopp, model, cat_features = walkthrough(classic_test=False, verbose=False)

omgång 1: https://www.atg.se/spel/2021-09-25/V75/
klickade på ANPASSA
anpassa klar - break
ant resultat 7
ant lopp 7


2021-09-26 01:40:56.428 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="421d24b7-23c0-4864-9ac6-06a2b7613330")>


Ant priser 7
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:40:56.821 WARNING root: 0


AVD 1 SOLVALLA 2140 AUTOSTART 

2021-09-26 01:40:57.511 WARNING root: 1


.

2021-09-26 01:40:58.209 WARNING root: 2


.

2021-09-26 01:40:58.993 WARNING root: 3


.

2021-09-26 01:40:59.758 WARNING root: 4


.

2021-09-26 01:41:00.518 WARNING root: 5


.

2021-09-26 01:41:01.314 WARNING root: 6


.

2021-09-26 01:41:02.176 WARNING root: 7


.

2021-09-26 01:41:02.987 WARNING root: 8


.

2021-09-26 01:41:03.794 WARNING root: 9


.

2021-09-26 01:41:04.633 WARNING root: 10


.

2021-09-26 01:41:05.478 WARNING root: 11


.

2021-09-26 01:41:06.352 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="492152ab-82bd-4b66-ba6c-82affe9090bc")>


.
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:41:06.788 WARNING root: 0


AVD 2 SOLVALLA 2140 AUTOSTART 

2021-09-26 01:41:07.657 WARNING root: 1


.

2021-09-26 01:41:08.543 WARNING root: 2


.

2021-09-26 01:41:09.450 WARNING root: 3


.

2021-09-26 01:41:10.373 WARNING root: 4


.

2021-09-26 01:41:11.317 WARNING root: 5


.

2021-09-26 01:41:12.249 WARNING root: 6


.

2021-09-26 01:41:13.215 WARNING root: 7


.

2021-09-26 01:41:14.189 WARNING root: 8


.

2021-09-26 01:41:15.175 WARNING root: 9


.

2021-09-26 01:41:16.184 WARNING root: 10


.

2021-09-26 01:41:17.197 WARNING root: 11


.

2021-09-26 01:41:18.221 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="0d784288-ee70-40a4-8560-4fc2a1a477ff")>


.
ant names,vodds,podds,rader,streck 15 15 15 15


2021-09-26 01:41:18.759 WARNING root: 0


AVD 3 SOLVALLA 2140 VOLTSTART 

2021-09-26 01:41:19.807 WARNING root: 1


.

2021-09-26 01:41:20.862 WARNING root: 2


.

2021-09-26 01:41:21.943 WARNING root: 3


.

2021-09-26 01:41:23.027 WARNING root: 4


.

2021-09-26 01:41:24.116 WARNING root: 5


.

2021-09-26 01:41:25.226 WARNING root: 6


.

2021-09-26 01:41:26.405 WARNING root: 7


.

2021-09-26 01:41:27.596 WARNING root: 8


.

2021-09-26 01:41:28.705 WARNING root: 9


.

2021-09-26 01:41:29.883 WARNING root: 10


.

2021-09-26 01:41:31.054 WARNING root: 11


.

2021-09-26 01:41:32.176 WARNING root: 12


.

2021-09-26 01:41:33.321 WARNING root: 13


.

2021-09-26 01:41:34.506 WARNING root: 14


.

2021-09-26 01:41:35.703 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="73bd57d9-066a-4676-b068-934e79570c38")>


.
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:41:36.181 WARNING root: 0


AVD 4 SOLVALLA 2140 AUTOSTART 

2021-09-26 01:41:37.379 WARNING root: 1


.

2021-09-26 01:41:38.563 WARNING root: 2


.

2021-09-26 01:41:39.805 WARNING root: 3


.

2021-09-26 01:41:41.078 WARNING root: 4


.

2021-09-26 01:41:42.361 WARNING root: 5


.

2021-09-26 01:41:43.658 WARNING root: 6


.

2021-09-26 01:41:45.154 WARNING root: 7


.

2021-09-26 01:41:46.434 WARNING root: 8


.

2021-09-26 01:41:47.685 WARNING root: 9


.

2021-09-26 01:41:48.947 WARNING root: 10


.

2021-09-26 01:41:50.237 WARNING root: 11


.

2021-09-26 01:41:51.551 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="ecdbee7c-3a85-4cb8-99d4-60392484c98f")>


.
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:41:52.021 WARNING root: 0


AVD 5 SOLVALLA 2140 AUTOSTART 

2021-09-26 01:41:53.374 WARNING root: 1


.

2021-09-26 01:41:54.715 WARNING root: 2


.

2021-09-26 01:41:56.065 WARNING root: 3


.

2021-09-26 01:41:57.438 WARNING root: 4


.

2021-09-26 01:41:59.049 WARNING root: 5


.

2021-09-26 01:42:00.594 WARNING root: 6


.

2021-09-26 01:42:02.254 WARNING root: 7


.

2021-09-26 01:42:03.592 WARNING root: 8


.

2021-09-26 01:42:04.738 WARNING root: 9


.

2021-09-26 01:42:05.928 WARNING root: 10


.

2021-09-26 01:42:07.157 WARNING root: 11


.

2021-09-26 01:42:08.338 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="7e492c2c-34bf-442e-9ef0-eb88ddfb5853")>


.
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:42:08.830 WARNING root: 0


AVD 6 SOLVALLA 2640 AUTOSTART 

2021-09-26 01:42:10.043 WARNING root: 1


.

2021-09-26 01:42:11.328 WARNING root: 2


.

2021-09-26 01:42:12.590 WARNING root: 3


.

2021-09-26 01:42:13.872 WARNING root: 4


.

2021-09-26 01:42:15.201 WARNING root: 5


.

2021-09-26 01:42:16.557 WARNING root: 6


.

2021-09-26 01:42:17.938 WARNING root: 7


.

2021-09-26 01:42:19.238 WARNING root: 8


.

2021-09-26 01:42:20.431 WARNING root: 9


.

2021-09-26 01:42:21.597 WARNING root: 10


.

2021-09-26 01:42:22.816 WARNING root: 11


.

2021-09-26 01:42:23.982 WARNING root: 2021-09-25 avd: <selenium.webdriver.remote.webelement.WebElement (session="aebbe3c8fc4b277f75b42179c85af781", element="cf62dfa6-6d4a-4b28-9fef-b90abb5af207")>


.
ant names,vodds,podds,rader,streck 12 12 12 12


2021-09-26 01:42:24.494 WARNING root: 0


AVD 7 SOLVALLA 2140 AUTOSTART 

2021-09-26 01:42:25.694 WARNING root: 1


.

2021-09-26 01:42:26.917 WARNING root: 2


.

2021-09-26 01:42:28.232 WARNING root: 3


.

2021-09-26 01:42:29.502 WARNING root: 4


.

2021-09-26 01:42:30.789 WARNING root: 5


.

2021-09-26 01:42:32.087 WARNING root: 6


.

2021-09-26 01:42:33.473 WARNING root: 7


.

2021-09-26 01:42:34.808 WARNING root: 8


.

2021-09-26 01:42:36.171 WARNING root: 9


.

2021-09-26 01:42:37.373 WARNING root: 10


.

2021-09-26 01:42:38.657 WARNING root: 11


..

det tog 103.92 sekunder
utdelning: 6502, 71, 0
startar Fixa mer
tog bort 3 strukna från 87 till 84
rensade totalt bort 3 hästar i städa_och_rensa. Från 87 till 84
shape med nya lopp (42006, 79)
shape med dubletter bort (42006, 79)
första datum i df = 2014-12-28
sista  datum i df = 2021-09-25
datum att lära från ['2021-09-25']
cat_features ['datum', 'bana', 'häst', 'kusk', 'kön', 'h1_kusk', 'h1_bana', 'h2_kusk', 'h2_bana', 'h3_kusk', 'h3_bana', 'h4_kusk', 'h4_bana', 'h5_kusk', 'h5_bana']

walk-iter 1 av 1 : best iteration None 	best score 0.75
spara model_senaste 2021-09-25


## Kör allt ovanför walkthrough
### Se till att "omg_att_spela_link.csv" är ifylld

# Kör cross validation

## init  - kör först allt t.o.m 'replace_NaN()' ovan

In [161]:
model = get_model().load_model('modeller/model_senaste')
dforg = pd.read_csv('all_data.csv')     
# print(df.columns)
df=remove_features(dforg.copy())
# df['avd']=dforg.avd
cat_features = list(df.loc[:,df.dtypes=='O'].columns)
df,_ = replace_NaN(df.copy(), cat_features=cat_features)    
y=df.plac
y=(y==1)*1
df.drop('plac',axis=1,inplace=True)

## cv

In [162]:

cv_pool = Pool(df,y,cat_features=cat_features)

params = {
         'use_best_model': True,
         'eval_metric' : 'AUC',
         "loss_function": "Logloss",
         'early_stopping_rounds': 100,
         'verbose': 50,
}

cv_score =cv(pool=cv_pool, 
   params=params, 
   dtrain=None, 
   iterations=2000, 
   num_boost_round=None,
   fold_count=5, 
   nfold=None,
   inverted=False,
   partition_random_seed=0,
   seed=2021, 
   shuffle=False, 
   logging_level=None, 
   stratified=True,
   as_pandas=True,
   type='TimeSeries')

0:	test: 0.6947994	best: 0.6947994 (0)
50:	test: 0.8108456	best: 0.8109098 (49)
100:	test: 0.8103996	best: 0.8117124 (77)
150:	test: 0.8093055	best: 0.8117124 (77)
Stopped by overfitting detector  (100 iterations wait)


In [163]:
cv_score

iterations  test-AUC-mean  test-AUC-std  test-Logloss-mean  \
0             0       0.694799      0.033239           0.659354   
1             1       0.729287      0.027954           0.627871   
2             2       0.746355      0.018905           0.600755   
3             3       0.753973      0.026538           0.573570   
4             4       0.759390      0.025517           0.549386   
..          ...            ...           ...                ...   
173         173       0.808774      0.005649           0.240215   
174         174       0.808782      0.005642           0.240209   
175         175       0.808759      0.005640           0.240220   
176         176       0.808591      0.005451           0.240288   
177         177       0.808552      0.005412           0.240294   

     test-Logloss-std  train-Logloss-mean  train-Logloss-std  
0            0.001047            0.659252           0.000977  
1            0.001831            0.627321           0.001671  
2            0.002615            0.600123           0.002735  
3            0.004307            0.572738           0.004244  
4            0.004098            0.548401           0.004036  
..                ...                 ...                ...  
173          0.001810            0.207509           0.006901  
174          0.001804            0.207431           0.006921  
175          0.001807            0.207294           0.006933  
176          0.001708            0.207185           0.007010  
177          0.001704            0.207044           0.007083  

[178 rows x 7 columns]

In [164]:
from IPython.display import display
print(df.datum.max())
display(cv_score[cv_score['test-Logloss-mean'].min() == cv_score['test-Logloss-mean']])
display(cv_score[cv_score['test-AUC-mean'].max() == cv_score['test-AUC-mean']])

2021-09-25


iterations  test-AUC-mean  test-AUC-std  test-Logloss-mean  \
142         142       0.809646      0.006655             0.2399   

     test-Logloss-std  train-Logloss-mean  train-Logloss-std  
142          0.002081            0.211149           0.005406

iterations  test-AUC-mean  test-AUC-std  test-Logloss-mean  \
77          77       0.811712      0.006353           0.244574   

    test-Logloss-std  train-Logloss-mean  train-Logloss-std  
77          0.003061            0.224611           0.002957

In [165]:
from sklearn.model_selection import train_test_split

df[['datum','avd','streck','häst','kusk']] = dforg[['datum','avd','streck','häst','kusk']]

# df.drop('datum',axis=1,inplace=True)
df.drop('avd',axis=1,inplace=True)
df.drop(['streck'],axis=1,inplace=True)
# df.drop(['häst','kusk'],axis=1,inplace=True)
cat_features = list(df.loc[:,df.dtypes=='O'].columns)
X_train,X_test,y_train,y_test = train_test_split(df,y,shuffle=False,)
cb=get_model(use_best=True)
cb.fit(X_train,y_train,eval_set= (X_test,y_test),early_stopping_rounds=200, cat_features=cat_features,verbose=100)

0:	learn: 0.6351497	test: 0.5750862	best: 0.5750862 (0)	total: 103ms	remaining: 5m 10s
100:	learn: 0.7109457	test: 0.6342845	best: 0.6401297 (18)	total: 14s	remaining: 6m 43s
200:	learn: 0.7375104	test: 0.6353957	best: 0.6431404 (156)	total: 28.2s	remaining: 6m 32s
300:	learn: 0.7529163	test: 0.6363518	best: 0.6431404 (156)	total: 41.2s	remaining: 6m 9s
Stopped by overfitting detector  (200 iterations wait)

bestTest = 0.6431404377
bestIteration = 156

Shrink model to first 157 iterations.


In [166]:
X_test[['datum','avd','vodds']] = dforg[['datum','avd','vodds']]
predict_prob = cb.predict_proba(X_test)

_,prob_score = proba_order_score(X_test ,y_test, predict_prob)

print('cb med ekipage',prob_score, cb.best_score_['validation']['AUC'])

cb med ekipage 4.052222222222222 0.707204980443889


C:\Users\peter\anaconda3\lib\site-packages\pandas\core\frame.py:3636: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


# FLAML (med och utan ekipage och streck)

In [297]:
def split_data(df_,train_from_proc=0,test_proc=0.25):
    # train_from_proc = where to start both train and test
    # test_proc = how much of the data is test
    df=df_.copy()
    alla_datum = df.datum.unique()
    train_from_datum = alla_datum[ int(len(alla_datum)*train_from_proc)]
    print(train_from_datum)
    X_test=None
    y_test=None
    test_from_datum=alla_datum[-1]
    if test_proc:
        selected_data = alla_datum[ alla_datum >= train_from_datum ]
        test_from_datum = selected_data[ int(len(selected_data)*(1-test_proc)) ]
        print(test_from_datum)
        X_test  = df[df.datum >= test_from_datum]
        y_test  = (X_test.plac==1)*1
        X_test  = X_test.drop('plac',axis=1)
        
    
    X_train = df[(df.datum >= train_from_datum) & (df.datum < test_from_datum) ]
    y_train = (X_train.plac==1)*1
    
    return X_train.drop('plac', axis=1), X_test, y_train, y_test


In [322]:
# prepare all data för flaml
dforg = pd.read_csv('all_data.csv')  

X_train, X_test, y_train, y_test= split_data(dforg,train_from_proc=0.2,test_proc=0.25)
X_train = remove_features(X_train)
X_test  = remove_features( X_test)

# X_train = X_train.drop('streck', axis=1)
# X_test  = X_test.drop ('streck', axis=1)
# X_train.drop('datum', axis=1, inplace=True)
# X_test.drop( 'datum', axis=1, inplace=True)
cat_features = list(X_train.select_dtypes('object').columns)
# X_train, X_test = replace_NaN(X_train.copy(),X_test=X_test.copy(), cat_features=cat_features) 
print(cat_features)
X_train.shape,X_test.shape

2016-06-18
2020-03-07
['datum', 'bana', 'häst', 'kusk', 'kön', 'h1_kusk', 'h1_bana', 'h2_kusk', 'h2_bana', 'h3_kusk', 'h3_bana', 'h4_kusk', 'h4_bana', 'h5_kusk', 'h5_bana']


((25189, 68), (8461, 68))

In [323]:
from flaml import AutoML 
cat_features = X_train.select_dtypes('object')
starting_points={'lgbm': {'n_estimators': 38,
  'num_leaves': 4,
  'min_child_samples': 2,
  'learning_rate': 0.19098448074739216,
  'log_max_bin': 7,
  'colsample_bytree': 0.8827412174089042,
  'reg_alpha': 0.004577823970660193,
  'reg_lambda': 0.03815584533462228},
 'rf': {'n_estimators': 33,
  'max_features': 0.3251674877768946,
  'max_leaves': 89,
  'criterion': 'entropy'},
 'catboost': {'early_stopping_rounds': 10,
  'learning_rate': 0.007511731949060241},
 'xgboost': {'n_estimators': 575,
  'max_leaves': 46,
  'min_child_weight': 1.032235057697502,
  'learning_rate': 0.013318439439138472,
  'subsample': 0.7908401179782586,
  'colsample_bylevel': 0.6924750037579576,
  'colsample_bytree': 0.7174828796230647,
  'reg_alpha': 0.15461500385937774,
  'reg_lambda': 0.6619886587472544},
 'extra_tree': {'n_estimators': 47,
  'max_features': 0.7934349565988307,
  'max_leaves': 213,
  'criterion': 'entropy'}}
flml_raw_parms={'task': 'classification','split_type':'time', 'metric':'roc_auc', 'starting_points': starting_points,'verbose':False,
        'time_budget':700, 'max_iter':50000000,'n_jobs':5, 'X_val': X_test, 'y_val':y_test,'early_stop':True, 'ensemble':True}

automl_raw = AutoML()
automl_raw.fit(X_train,y_train, **flml_raw_parms)

2021-09-28 22:41:17.537 WARNING flaml.searcher.blendsearch: No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'.
C:\Users\peter\AppData\Roaming\Python\Python38\site-packages\sklearn\linear_model\_sag.py:328: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn("The max_iter was reached which means "


In [324]:
flm_raw_train_pred= automl_raw.predict_proba(X_train)
flm_raw_test_pred = automl_raw.predict_proba(X_test)

X_test_raw = X_test.copy()
X_test_raw[['datum','avd','vodds']] = dforg[['datum','avd','vodds']]
_,prob_score = proba_order_score(X_test_raw,y_test, flm_raw_test_pred)

print('timeserie, datum,häst, kusk', prob_score, 1-automl_raw.best_loss)
# X_test.columns

timeserie, datum,häst, kusk 3.0824175824175826 0.8220287891340522


timeserie  0.3 0.25, datum, häst, kusk 3.720565149136578  0.7213763318649257   best   
timeserie  0.4 0.25, datum, häst, kusk 3.7362637362637363 0.7214144007762124  
timeserie, 0.2 0.25, datum, häst, kusk 3.760989010989011  0.72561915325073230  
timeserie, 0.1 0.25  datum, häst, kusk 3.8180708180708183 0.726597977829505    
timeserie, 0.5 0.25, datum, häst, kusk 3.936263736263736  0.7216626969090024  
timeserie, 0.3 0.25, datum, häst, kusk, streck 3.06436420722135   0.8232840226857013 best      
timeserie, 0.4 0.25, datum, häst, kusk, streck 3.1483516483516483 0.8169106155467452  
timeserie, 0.2 0.25, datum, häst, kusk, streck 3.0824175824175826 0.8220287891340522

## Final FLML model

In [269]:
X_train, X_test, y_train, y_test= split_data(remove_features(dforg),train_from_proc=0.3,test_proc=None)

# df.drop('streck', axis=1, inplace=True)
# df = create_ekipage(df)
# df.drop('datum', axis=1, inplace=True)

# X_train, X_test = replace_NaN(X_train.copy(),X_test=X_test.copy(), cat_features=cat_features) 
# X_train.drop('steck',axis=1,inplace=True)
# X_train, _ = replace_NaN(X_train.copy(), cat_features=cat_features) 
cat_features = list(X_train.select_dtypes('object').columns)
print(cat_features)
automl = AutoML()
flml_parms={'task': 'classification','split_type':'time', 'metric':'roc_auc','starting_points': starting_points, 'verbose':False,
        'time_budget':1000, 'max_iter':90000000,'n_jobs':5, 'early_stop':True, 'ensemble':True}

2017-02-04
['datum', 'bana', 'häst', 'kusk', 'kön', 'h1_kusk', 'h1_bana', 'h2_kusk', 'h2_bana', 'h3_kusk', 'h3_bana', 'h4_kusk', 'h4_bana', 'h5_kusk', 'h5_bana']


In [271]:
automl.fit(X_train, y_train, **flml_parms)
1-automl.best_loss

0.8190833786268518

In [177]:
import pickle
filename = 'modeller\\FLAML._model.sav'
pickle.dump(automl_raw, open(filename, 'wb'))